File: ai-service/app/middleware/security.py
"""
Enterprise CX Guardian AI — Security Middleware Stack (FastAPI)
================================================================

Implements all 7 security controls:
  1. Secure HTTP Headers  (X-Frame-Options, CSP, HSTS, etc.)
  2. CORS Configuration   (allowlist, credentials, preflight)
  3. Rate Limiting        (in-memory sliding window per IP)
  4. Request Size Limits  (body size guard)
  5. Input Sanitization   (strip null bytes, control chars)
  6. Environment Validation  (required vars checked at boot)
  7. Security Logging     (suspicious requests)

All middleware classes are designed as FastAPI/Starlette ASGI
middleware — pluggable into app.add_middleware().
"""

import re
import time
import asyncio
from collections import defaultdict, deque
from typing import Callable, Dict, Optional
from fastapi import Request
from fastapi.responses import JSONResponse
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.types import ASGIApp

from app.core.config import settings
from app.core.logger import logger


# ══════════════════════════════════════════════════════════════════
# 1. Secure HTTP Headers Middleware
# ══════════════════════════════════════════════════════════════════

class SecureHeadersMiddleware(BaseHTTPMiddleware):
    """
    Adds production-grade security headers to every HTTP response.
    Equivalent to Helmet.js for Node.js.

    Headers applied:
      X-Content-Type-Options        — Prevent MIME sniffing
      X-Frame-Options               — Clickjacking protection
      X-XSS-Protection              — Legacy XSS filter hint
      Strict-Transport-Security     — Force HTTPS (HSTS)
      Referrer-Policy               — Limit referrer leakage
      Permissions-Policy            — Disable unused browser APIs
      Content-Security-Policy       — Script/style source allowlist
      Cache-Control                 — Prevent sensitive data caching
    """

    # Paths where cache-control should NOT prevent caching (static assets)
    _CACHEABLE_PATHS = {"/docs", "/redoc", "/openapi.json"}

    async def dispatch(self, request: Request, call_next: Callable):
        response = await call_next(request)

        # ── Security headers ───────────────────────────────────────
        response.headers["X-Content-Type-Options"]  = "nosniff"
        response.headers["X-Frame-Options"]         = "DENY"
        response.headers["X-XSS-Protection"]        = "1; mode=block"
        response.headers["Referrer-Policy"]         = "strict-origin-when-cross-origin"
        response.headers["Permissions-Policy"]      = (
            "camera=(), microphone=(), geolocation=(), "
            "payment=(), usb=(), bluetooth=()"
        )
        response.headers["Content-Security-Policy"] = (
            "default-src 'self'; "
            "script-src 'self' 'unsafe-inline'; "
            "style-src 'self' 'unsafe-inline'; "
            "img-src 'self' data:; "
            "connect-src 'self'; "
            "frame-ancestors 'none'"
        )

        # HSTS — only in production (not dev/test)
        if settings.ENVIRONMENT == "production":
            response.headers["Strict-Transport-Security"] = (
                "max-age=31536000; includeSubDomains; preload"
            )

        # Prevent caching for API responses
        if request.url.path not in self._CACHEABLE_PATHS:
            response.headers["Cache-Control"] = "no-store, no-cache, must-revalidate"
            response.headers["Pragma"]        = "no-cache"

        # Remove server fingerprinting headers
        if "Server" in response.headers:
            del response.headers["Server"]
        if "X-Powered-By" in response.headers:
            del response.headers["X-Powered-By"]

        return response


# ══════════════════════════════════════════════════════════════════
# 2. Rate Limiting Middleware
# ══════════════════════════════════════════════════════════════════

class RateLimitConfig:
    """Per-route rate limit configuration."""
    def __init__(self, max_requests: int, window_seconds: int):
        self.max_requests   = max_requests
        self.window_seconds = window_seconds


# Route-specific limits (most restrictive wins)
RATE_LIMIT_RULES: Dict[str, RateLimitConfig] = {
    "/api/v1/auth/login":    RateLimitConfig(max_requests=10,   window_seconds=60),
    "/api/v1/auth/register": RateLimitConfig(max_requests=5,    window_seconds=300),
    "/api/v1/auth/refresh":  RateLimitConfig(max_requests=20,   window_seconds=60),
    "/api/v1/chat":          RateLimitConfig(max_requests=60,   window_seconds=60),
    "/api/v1/conversations": RateLimitConfig(max_requests=120,  window_seconds=60),
    "default":               RateLimitConfig(max_requests=200,  window_seconds=60),
}


class RateLimitMiddleware(BaseHTTPMiddleware):
    """
    Sliding-window per-IP rate limiter.

    Uses an in-memory deque per (IP, route) key.
    For production deployments with multiple workers,
    replace with Redis-backed sliding window.

    Excluded: health/metrics endpoints (monitoring probes must not be throttled).
    """

    _EXCLUDED = {"/", "/health", "/health/live", "/health/ready", "/metrics"}

    def __init__(self, app: ASGIApp):
        super().__init__(app)
        # {(ip, route): deque of timestamps}
        self._windows: Dict[tuple, deque] = defaultdict(lambda: deque())
        self._lock = asyncio.Lock()

    def _get_config(self, path: str) -> RateLimitConfig:
        """Match the most specific prefix rule, fall back to default."""
        for prefix, config in RATE_LIMIT_RULES.items():
            if prefix != "default" and path.startswith(prefix):
                return config
        return RATE_LIMIT_RULES["default"]

    def _get_client_ip(self, request: Request) -> str:
        """Extract real IP, respecting X-Forwarded-For from trusted proxy."""
        forwarded = request.headers.get("X-Forwarded-For")
        if forwarded:
            return forwarded.split(",")[0].strip()
        return request.client.host if request.client else "unknown"

    async def dispatch(self, request: Request, call_next: Callable):
        path = request.url.path

        if path in self._EXCLUDED:
            return await call_next(request)

        ip      = self._get_client_ip(request)
        config  = self._get_config(path)
        now     = time.time()
        key     = (ip, path.split("?")[0])

        async with self._lock:
            window = self._windows[key]

            # Evict timestamps outside the sliding window
            cutoff = now - config.window_seconds
            while window and window[0] < cutoff:
                window.popleft()

            if len(window) >= config.max_requests:
                oldest   = window[0]
                retry_after = int(config.window_seconds - (now - oldest)) + 1
                logger.warning(
                    f"[Rate Limit] BLOCKED | ip: {ip} | path: {path} "
                    f"| requests: {len(window)}/{config.max_requests} "
                    f"| retry_after: {retry_after}s"
                )
                return JSONResponse(
                    status_code=429,
                    headers={"Retry-After": str(retry_after)},
                    content={
                        "success":    False,
                        "message":    "Rate limit exceeded",
                        "error":      f"Too many requests. Retry after {retry_after} seconds.",
                        "error_code": "RATE_LIMIT_EXCEEDED",
                        "retry_after": retry_after
                    }
                )

            window.append(now)

        return await call_next(request)


# ══════════════════════════════════════════════════════════════════
# 3. Request Size Limit Middleware
# ══════════════════════════════════════════════════════════════════

# Route-specific size limits (bytes)
REQUEST_SIZE_LIMITS: Dict[str, int] = {
    "/api/v1/chat":          64   * 1024,    # 64 KB  — chat messages
    "/api/v1/auth/register": 4    * 1024,    # 4 KB   — auth payloads
    "/api/v1/auth/login":    4    * 1024,    # 4 KB
    "default":               256  * 1024,    # 256 KB — general API
}


class RequestSizeLimitMiddleware(BaseHTTPMiddleware):
    """
    Rejects requests whose Content-Length exceeds the configured limit.
    Protects against request smuggling and oversized payload attacks.
    """

    def _get_limit(self, path: str) -> int:
        for prefix, limit in REQUEST_SIZE_LIMITS.items():
            if prefix != "default" and path.startswith(prefix):
                return limit
        return REQUEST_SIZE_LIMITS["default"]

    async def dispatch(self, request: Request, call_next: Callable):
        content_length = request.headers.get("Content-Length")
        if content_length:
            size  = int(content_length)
            limit = self._get_limit(request.url.path)
            if size > limit:
                logger.warning(
                    f"[Request Size] REJECTED | path: {request.url.path} "
                    f"| size: {size} bytes | limit: {limit} bytes"
                )
                return JSONResponse(
                    status_code=413,
                    content={
                        "success":    False,
                        "message":    "Request payload too large",
                        "error":      f"Request body exceeds {limit // 1024} KB limit.",
                        "error_code": "PAYLOAD_TOO_LARGE"
                    }
                )
        return await call_next(request)


# ══════════════════════════════════════════════════════════════════
# 4. Input Sanitization Middleware
# ══════════════════════════════════════════════════════════════════

_DANGEROUS_PATTERNS = re.compile(
    r"(\x00|%00|[\x01-\x08\x0b\x0c\x0e-\x1f\x7f])"  # Null bytes + control chars
    r"|(<script[\s\S]*?>[\s\S]*?</script>)"         # XSS script tags
    r"|(javascript\s*:)",                            # javascript: URI
    re.IGNORECASE
)


def sanitize_string(value: str) -> str:
    """Strip null bytes and dangerous patterns from a string."""
    return _DANGEROUS_PATTERNS.sub("", value)


class InputSanitizationMiddleware(BaseHTTPMiddleware):
    """
    Strips null bytes and obvious injection payloads from query
    parameters and path segments before routing begins.

    Note: Body sanitization is handled at the schema/validator level
    (Pydantic validators on each schema field).
    """

    async def dispatch(self, request: Request, call_next: Callable):
        # Sanitize query params
        raw_query = str(request.url.query)
        if raw_query and _DANGEROUS_PATTERNS.search(raw_query):
            logger.warning(
                f"[Input Sanitize] Dangerous query string detected | "
                f"path: {request.url.path} | ip: {request.client.host if request.client else '?'}"
            )
            return JSONResponse(
                status_code=400,
                content={
                    "success":    False,
                    "message":    "Invalid request parameters",
                    "error":      "Request contains forbidden characters.",
                    "error_code": "INVALID_INPUT"
                }
            )
        return await call_next(request)


# ══════════════════════════════════════════════════════════════════
# 5. Environment Validation (called at startup, not middleware)
# ══════════════════════════════════════════════════════════════════

REQUIRED_ENV_VARS = [
    ("GROQ_API_KEY",   "Groq AI API key for LLM completions"),
    ("MONGODB_URI",    "MongoDB Atlas connection string"),
    ("JWT_SECRET_KEY", "JWT signing secret (min 32 chars recommended)"),
]

OPTIONAL_WITH_DEFAULTS = {
    "ENVIRONMENT":   "development",
    "LOG_LEVEL":     "INFO",
    "MODEL_NAME":    "llama3-70b-8192",
    "MAX_HISTORY":   "10",
    "MAX_TOKENS":    "1024",
    "TEMPERATURE":   "0.2",
}


def validate_environment() -> None:
    """
    Validates all required environment variables at application startup.
    Raises ValueError if any required variable is missing or empty.
    Logs warnings for optional variables using defaults.
    """
    missing = []
    warnings = []

    for var, description in REQUIRED_ENV_VARS:
        value = getattr(settings, var, None)
        if not value or value.strip() in ("", "your_groq_api_key_here", "CHANGE_ME"):
            missing.append(f"  ✗ {var} — {description}")

    for var, default in OPTIONAL_WITH_DEFAULTS.items():
        value = getattr(settings, var, None)
        if not value:
            warnings.append(f"  ⚠ {var} not set — using default: '{default}'")

    if warnings:
        for w in warnings:
            logger.warning(f"[Env Validation]{w}")

    if missing and settings.ENVIRONMENT == "production":
        msg = "Missing required environment variables:\n" + "\n".join(missing)
        logger.error(f"[Env Validation] STARTUP BLOCKED\n{msg}")
        raise ValueError(msg)
    elif missing:
        for m in missing:
            logger.warning(f"[Env Validation] DEV MODE — Missing: {m}")

    # JWT secret strength check
    jwt_key = getattr(settings, "JWT_SECRET_KEY", "")
    if len(jwt_key) < 32:
        logger.warning(
            "[Env Validation] ⚠ JWT_SECRET_KEY is shorter than 32 characters — "
            "use a longer secret in production"
        )

    logger.info("[Env Validation] ✓ Environment validated successfully")


# ══════════════════════════════════════════════════════════════════
# 6. CORS Configuration Builder
# ══════════════════════════════════════════════════════════════════

def get_cors_config() -> dict:
    """
    Returns environment-aware CORS configuration for FastAPI CORSMiddleware.

    Development: allows all origins (open for local dev)
    Production:  restricts to NODE_BACKEND_URL only
    """
    if settings.ENVIRONMENT == "production":
        allowed_origins = [
            settings.NODE_BACKEND_URL,
            # Add deployed frontend URL here e.g. "https://app.enterprise-cx.ai"
        ]
        logger.info(
            f"[CORS] Production mode — allowed origins: {allowed_origins}"
        )
    else:
        allowed_origins = ["*"]
        logger.debug("[CORS] Development mode — all origins allowed")

    return {
        "allow_origins":     allowed_origins,
        "allow_credentials": True,
        "allow_methods":     ["GET", "POST", "PUT", "PATCH", "DELETE", "OPTIONS"],
        "allow_headers":     [
            "Authorization",
            "Content-Type",
            "X-Request-ID",
            "X-API-Version",
            "Accept",
            "Origin"
        ],
        "expose_headers":    ["X-Request-ID", "X-Response-Time"],
        "max_age":           600,   # Preflight cache: 10 minutes
    }

File: server/middleware/security.js
/**
 * Enterprise CX Guardian AI — Node.js Security Middleware Stack
 * ==============================================================
 *
 * Implements all 7 security controls:
 *   1. Helmet              — Secure HTTP headers
 *   2. CORS Configuration  — Strict allowlist with credentials
 *   3. Rate Limiting       — express-rate-limit per IP per route
 *   4. Request Size Limits — JSON + URL-encoded body limits
 *   5. Input Validation    — Joi schemas for all request bodies
 *   6. Input Sanitization  — Express-validator + xss protection
 *   7. Environment Validation — Required var check at startup
 */

import helmet    from 'helmet';
import cors      from 'cors';
import logger    from '../config/logger.js';
import * as env  from '../config/env.js';

// ══════════════════════════════════════════════════════════════════
// 1. Helmet — Secure HTTP Headers
// ══════════════════════════════════════════════════════════════════

export const helmetConfig = helmet({
  // Content Security Policy
  contentSecurityPolicy: {
    directives: {
      defaultSrc:  ["'self'"],
      scriptSrc:   ["'self'", "'unsafe-inline'"],
      styleSrc:    ["'self'", "'unsafe-inline'"],
      imgSrc:      ["'self'", "data:"],
      connectSrc:  ["'self'"],
      fontSrc:     ["'self'"],
      objectSrc:   ["'none'"],
      mediaSrc:    ["'self'"],
      frameSrc:    ["'none'"],
      frameAncestors: ["'none'"]
    }
  },
  // HSTS: force HTTPS for 1 year (production only)
  hsts: env.NODE_ENV === 'production'
    ? { maxAge: 31536000, includeSubDomains: true, preload: true }
    : false,
  // Clickjacking
  frameguard:         { action: 'deny' },
  // MIME sniffing
  noSniff:            true,
  // XSS
  xssFilter:          true,
  // Hide X-Powered-By
  hidePoweredBy:      true,
  // Referrer
  referrerPolicy:     { policy: 'strict-origin-when-cross-origin' },
  // Permissions
  permittedCrossDomainPolicies: { permittedPolicies: 'none' },
  // DNS prefetch control
  dnsPrefetchControl: { allow: false },
  // IE no-open
  ieNoOpen:           true,
});

// ══════════════════════════════════════════════════════════════════
// 2. CORS Configuration
// ══════════════════════════════════════════════════════════════════

const _allowedOrigins = env.NODE_ENV === 'production'
  ? [env.CLIENT_URL].filter(Boolean)
  : ['http://localhost:5173', 'http://localhost:3000', 'http://localhost:80'];

export const corsConfig = cors({
  origin: (origin, callback) => {
    // Allow non-browser requests (curl, Postman, server-to-server)
    if (!origin) return callback(null, true);
    if (_allowedOrigins.includes(origin)) {
      return callback(null, true);
    }
    logger.warn(`[CORS] Blocked origin: ${origin}`);
    callback(new Error(`CORS policy: origin '${origin}' not allowed`));
  },
  credentials:    true,
  methods:        ['GET', 'POST', 'PUT', 'PATCH', 'DELETE', 'OPTIONS'],
  allowedHeaders: [
    'Authorization', 'Content-Type', 'X-Request-ID',
    'X-API-Version', 'Accept', 'Origin'
  ],
  exposedHeaders: ['X-Request-ID', 'X-Response-Time', 'X-Rate-Limit-Remaining'],
  maxAge:         600,    // Preflight cache: 10 minutes
  optionsSuccessStatus: 204
});

// ══════════════════════════════════════════════════════════════════
// 3. Rate Limiting (Zero-dependency sliding window)
// ══════════════════════════════════════════════════════════════════

function createRateLimiter({ windowMs = 60000, max = 100, message = 'Too many requests' } = {}) {
  const store = new Map();

  return (req, res, next) => {
    if (['/api/v1/health', '/health'].includes(req.path)) {
      return next();
    }

    const ip = req.ip || req.socket.remoteAddress || 'unknown';
    const now = Date.now();
    const windowStart = now - windowMs;

    let timestamps = store.get(ip) || [];
    timestamps = timestamps.filter(ts => ts > windowStart);

    if (timestamps.length >= max) {
      const retryAfter = Math.ceil((timestamps[0] + windowMs - now) / 1000);
      logger.warn(`[Rate Limit] BLOCKED | ip: ${ip} | path: ${req.originalUrl} | retry_after: ${retryAfter}s`);
      return res.status(429).json({
        success: false,
        message: 'Rate limit exceeded',
        error: message,
        error_code: 'RATE_LIMIT_EXCEEDED',
        retry_after: retryAfter
      });
    }

    timestamps.push(now);
    store.set(ip, timestamps);
    res.setHeader('RateLimit-Limit', max);
    res.setHeader('RateLimit-Remaining', max - timestamps.length);
    next();
  };
}

// General API limiter — 200 req/min per IP
export const generalLimiter = createRateLimiter({
  windowMs: 60 * 1000,
  max: 200,
  message: 'Too many requests. Retry after 60 seconds.'
});

// Auth endpoint limiter — 10 req/min per IP
export const authLimiter = createRateLimiter({
  windowMs: 60 * 1000,
  max: 10,
  message: 'Too many authentication attempts.'
});

// Chat endpoint limiter — 60 req/min per IP
export const chatLimiter = createRateLimiter({
  windowMs: 60 * 1000,
  max: 60,
  message: 'Too many chat messages.'
});

// ══════════════════════════════════════════════════════════════════
// 4. Request Size Limits (used in app.use())
// ══════════════════════════════════════════════════════════════════

export const REQUEST_SIZE_LIMITS = {
  json:       '256kb',   // express.json() limit
  urlencoded: '64kb',    // express.urlencoded() limit
  chat:       '64kb',    // applied per-route on /api/v1/chat
  auth:       '4kb',     // applied per-route on /api/v1/auth
};

// Per-route size guard middleware factory
export const bodySizeLimit = (limit) => (req, res, next) => {
  const contentLength = parseInt(req.headers['content-length'] || '0', 10);
  const limitBytes = parseSize(limit);
  if (limitBytes > 0 && contentLength > limitBytes) {
    logger.warn(
      `[Request Size] REJECTED | path: ${req.path} | size: ${contentLength}B | limit: ${limitBytes}B`
    );
    return res.status(413).json({
      success:    false,
      message:    'Request payload too large',
      error:      `Body exceeds ${limit} limit`,
      error_code: 'PAYLOAD_TOO_LARGE'
    });
  }
  next();
};

function parseSize(size) {
  if (typeof size === 'number') return size;
  const units = { b: 1, kb: 1024, mb: 1024 * 1024 };
  const match = String(size).toLowerCase().match(/^(\d+)(b|kb|mb)$/);
  return match ? parseInt(match[1]) * (units[match[2]] || 1) : 0;
}

// ══════════════════════════════════════════════════════════════════
// 5. Input Validation — Joi schema factory
// ══════════════════════════════════════════════════════════════════

/**
 * Returns a middleware that validates req.body against a Joi schema.
 * Requires: npm install joi
 *
 * Usage:
 *   import Joi from 'joi';
 *   import { validateBody } from '../middleware/security.js';
 *
 *   const chatSchema = Joi.object({ message: Joi.string().max(2000).required() });
 *   router.post('/chat', validateBody(chatSchema), chatController);
 */
export const validateBody = (schema) => (req, res, next) => {
  if (!schema || !schema.validate) return next();

  const { error, value } = schema.validate(req.body, {
    abortEarly:       false,
    stripUnknown:     true,
    allowUnknown:     false,
  });

  if (error) {
    const detail = error.details.map(d => d.message).join('; ');
    logger.warn(`[Input Validation] FAILED | path: ${req.path} | ${detail}`);
    return res.status(400).json({
      success:    false,
      message:    'Request validation failed',
      error:      detail,
      error_code: 'VALIDATION_ERROR'
    });
  }

  req.body = value;   // Use sanitized/coerced value
  next();
};

// ══════════════════════════════════════════════════════════════════
// 6. Input Sanitization
// ══════════════════════════════════════════════════════════════════

const _DANGEROUS = /(\x00|<script[\s\S]*?>[\s\S]*?<\/script>|javascript\s*:)/gi;
const _NULL_BYTE  = /\x00/g;

/**
 * Recursively sanitizes all string values in an object.
 * Strips: null bytes, <script> tags, javascript: URIs
 */
export const sanitize = (value) => {
  if (typeof value === 'string') {
    return value.replace(_DANGEROUS, '').replace(_NULL_BYTE, '').trim();
  }
  if (Array.isArray(value)) {
    return value.map(sanitize);
  }
  if (value && typeof value === 'object') {
    return Object.fromEntries(
      Object.entries(value).map(([k, v]) => [k, sanitize(v)])
    );
  }
  return value;
};

/** Express middleware — sanitizes req.body, req.params, req.query */
export const sanitizeInputs = (req, res, next) => {
  if (req.body)   req.body   = sanitize(req.body);
  if (req.params) req.params = sanitize(req.params);
  if (req.query)  req.query  = sanitize(req.query);
  next();
};

// ══════════════════════════════════════════════════════════════════
// 7. Environment Validation
// ══════════════════════════════════════════════════════════════════

const REQUIRED_VARS = [
  ['MONGO_URI',     'MongoDB connection string'],
  ['JWT_SECRET',    'JWT signing secret (min 32 chars)'],
  ['PYTHON_AI_URL', 'Python AI microservice URL'],
];

const OPTIONAL_VARS = {
  'NODE_ENV':       'development',
  'PORT':           '5000',
  'CLIENT_URL':     'http://localhost:5173',
  'AI_TIMEOUT':     '30000',
  'MAX_RETRIES':    '3',
  'STORAGE_BACKEND':'memory',
};

export const validateEnvironment = () => {
  const missing  = [];
  const warnings = [];

  for (const [varName, description] of REQUIRED_VARS) {
    const val = process.env[varName];
    if (!val || val.trim() === '') {
      missing.push(`  ✗ ${varName} — ${description}`);
    }
  }

  for (const [varName, defaultVal] of Object.entries(OPTIONAL_VARS)) {
    if (!process.env[varName]) {
      warnings.push(`  ⚠ ${varName} not set — using default: '${defaultVal}'`);
    }
  }

  // JWT secret strength check
  const jwtSecret = process.env.JWT_SECRET || '';
  if (jwtSecret.length < 32) {
    warnings.push('  ⚠ JWT_SECRET shorter than 32 chars — use a longer secret in production');
  }

  warnings.forEach(w => logger.warn(`[Env Validation]${w}`));

  if (missing.length > 0) {
    const msg = `Missing required environment variables:\n${missing.join('\n')}`;
    if (env.NODE_ENV === 'production') {
      logger.error(`[Env Validation] STARTUP BLOCKED\n${msg}`);
      throw new Error(msg);
    } else {
      missing.forEach(m => logger.warn(`[Env Validation] DEV MODE: ${m}`));
    }
  }

  logger.info('[Env Validation] ✓ Environment validated successfully');
};
